### ISSUE TITLE
Array.map() produces unexpected error

### ISSUE THREAD

**Issue by KROT47** [facebook/flow#4294]
```js
/* @flow */

const arr = [{
  key1: 1,
}, {
  key2: 2,
}];

var result = arr.map( obj => Object.keys( obj ).map( key => ({
  [ key ]: obj[ key ] || undefined,
})));

arr.map( obj => ({
  a: obj.key1,
  // but here error goes away with "|| undefined"
  b: obj.key2 || undefined,
}));
```

[try flow example](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVBjOA7AzgFzAEMAnEsAXjAG0BvVMMAawFMBPARgC4wOAaVAF8+Yeo1ZsATD0kDBAXQDc6AG6kwJFrgCuMQlVIkAdAFsiABwAUYOACMAVpQB8YAPIOWGfEYm5rdxwBKUwtrCWcwSzEaZnYweR4A6li2eLAAH3SwbWwAExYoAEtsFly5QIrlVEMQqxsHCKiGYkSHH3Z+ZuBgMFttQgALFk0wYZI4cgBzOC1iBCJUhEL8AbAAIkzsvILi0rXm21b7dqkMrJz8opKyoUqgA)

Probably similar to [#4172](https://github.com/facebook/flow/issues/4172)

**Comment by TrySound** [comment]
In the second example you are accessing possibly not existing field and you need to unsure it exists. So `|| undefined` is explicit way to do it.

In first I guess there's a problem with computed properties (`obj[key]`).

**Comment by KROT47** [comment]
Agreed, so there is a problem, since I explicitly tell in first example that undefined is ok and since `obj` is an object literal ( not primitive type ), like I do in second example.

**Comment by lll000111** [comment]
This would be more interesting with types (annotations). I think it is okay and part of the philosophy of these type systems to bark at you when you mix different objects (different types) in one array. If you want to do that you add annotations to tell the type system what exactly you expect from an object in that array.

Type systems _do_ limit what you can do, you can only use a subset of the Javascript that you could write without them. I don't think this is a bug unless you add Flow types and show that stuff that is supposed to work doesn't.

**Comment by SamChou19815** [comment]
We will now error in the Object.keys map, which is expected.
